x11rb-async/src/rust_connection/shared_state.rs

```diff
@@ -181,3 +200,16 @@ impl PacketReader {
         Ok(())
     }
 }
+
+#[derive(Debug)]
+pub(super) struct BreakOnDrop<S>(pub(super) Arc<SharedState<S>>);
```

Comment:
I feel like we could probably avoid using a clone here, and do this instead:

```rust
async fn drive(&self, ...) {
    let _drop = CallOnDrop(|| {
        self.driver_broken.store(...);
        ...
    });

    ...
}

struct CallOnDrop<F: Fn()>(F);

impl<F: Fn()> Drop for CallOnDrop<F> {
    fn drop(&mut self) {
        (self.0)();
    }
}
```

This is a more common pattern, I think.

Reply:
Yeah, but that does not work with an async function. If I drop the future before it was ever polled even once, then the `CallOnDrop` would not yet have been created.

Basically: The following program does not print `Hello world`:
```rust
async fn foo() {
    println!("Hello, world!");
}

fn main() {
    drop(foo());
}
```

Reply:
Since this is apparently a special "special case", I added a test for this. This test now hangs when applying the following diff (which hopefully models your suggestion):
```diff
diff --git a/x11rb-async/src/rust_connection/mod.rs b/x11rb-async/src/rust_connection/mod.rs
index eae9902..fcd6e61 100644
--- a/x11rb-async/src/rust_connection/mod.rs
+++ b/x11rb-async/src/rust_connection/mod.rs
@@ -220,8 +220,7 @@ impl<S: Stream + Send + Sync> RustConnection<S> {
             // When the following object is dropped, it marks the connection as broken. This
             // mechanism fires when the future from drive() is dropped. We need to create it
             // outside of the async block in case that future is never polled.
-            let break_on_drop = shared_state::BreakOnDrop(shared.clone());
-            async move { shared.drive(break_on_drop).await }
+            async move { shared.drive().await }
         };
 
         Ok((
diff --git a/x11rb-async/src/rust_connection/shared_state.rs b/x11rb-async/src/rust_connection/shared_state.rs
index 5796131..286ae77 100644
--- a/x11rb-async/src/rust_connection/shared_state.rs
+++ b/x11rb-async/src/rust_connection/shared_state.rs
@@ -89,8 +89,15 @@ impl<S: Stream> SharedState<S> {
     /// Read incoming packets from the stream and put them into the inner connection.
     pub(super) async fn drive(
         &self,
-        _break_on_drop: BreakOnDrop<S>,
     ) -> Result<Infallible, ConnectionError> {
+        let _drop = CallOnDrop(|| {
+            // Mark the connection as broken
+            self.driver_dropped.store(true, Ordering::SeqCst);
+
+            // Wake up everyone that might be waiting
+            self.new_input.notify_additional(std::usize::MAX);
+        });
+
         let mut packet_reader = PacketReader {
             read_buffer: vec![0; 4096].into_boxed_slice(),
             inner: ProtoPacketReader::new(),
@@ -213,3 +220,11 @@ impl<S> Drop for BreakOnDrop<S> {
         self.0.new_input.notify_additional(std::usize::MAX);
     }
 }
+
+struct CallOnDrop<F: Fn()>(F);
+
+impl<F: Fn()> Drop for CallOnDrop<F> {
+    fn drop(&mut self) {
+        (self.0)();
+    }
+}
```

Reply:
Ah, I see. Then that makes sense.